### PR TITLE
Add validation and optimise objects.BulkDelete

### PR DESCRIPTION
**Pull request by pierreprinetti**
This PR makes the following changes to `objects.BulkDelete`:
* validate that the container name does not contain `/`;
* validate that all provided objects are non-empty;
* optimise how the request body is built.

Note that the function mentions urlencoding in a comment, but there is no urlencoding. I have removed the comment and kept the behaviour. We might want to consider add urlencoding in a follow-up.

---

WRT the optimisation:

Use `bytes.Buffer` instead of repeatedly copying strings to compose the request body for `objects.BulkDelete`.

This PR adds, and in a separate commit, removes the benchmark tests. To execute the benchmark:

```shell
git reset --hard f282ec49
cd openstack/objectstorage/v1/objects/testing
go test -bench=.
```

Locally I get a marginal gain when bulk-deleting only a few objects, and
good gains (to around 3/4 of the time per execution) with large batches:

```plaintext
goos: linux
goarch: amd64
pkg: github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects/testing
cpu: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
BenchmarkBulkDelete-8                                246           4536244 ns/op
BenchmarkBulkDeleteFewerObjects-8                  11943             97338 ns/op
BenchmarkBulkDeleteWithBuffer-8                      303           3463102 ns/op
BenchmarkBulkDeleteWithBufferFewerObjects-8        11587             92153 ns/op
PASS
ok      github.com/gophercloud/gophercloud/openstack/objectstorage/v1/objects/testing   19.943s
```

### CODE REVIEW

**Comment by kayrus** [comment]
@pierreprinetti I wonder whether it will work with the new versioning system
https://github.com/gophercloud/gophercloud/blob/master/acceptance/openstack/objectstorage/v1/versioning_test.go

**Comment by kayrus** [comment]
@pierreprinetti I don't know. Just thinking out loud. It may happen that this function consumers may face 409 on objects or container deletion :\

File: openstack/objectstorage/v1/objects/requests.go [hunk 1]
```diff
@@ -576,15 +576,25 @@ func CreateTempURL(c *gophercloud.ServiceClient, containerName, objectName strin
 
 // BulkDelete is a function that bulk deletes objects.
```
Comment (stephenfin) [inline openstack/objectstorage/v1/objects/requests.go:577]: Unrelated to this change, but I wonder if we should mention the (configurable) limit of 10,000 objects per request?

**Review by stephenfin — APPROVED** [review]
Looks sane to me

File: openstack/objectstorage/v1/objects/requests.go (hunk 1 above)
Comment (pierreprinetti) [reply openstack/objectstorage/v1/objects/requests.go:577]: I wasn't able to find a reference to a limit [in the API docs](https://docs.openstack.org/api-ref/object-store/#delete-object). Did you find that in Swift or Ceph?

**Review by EmilienM — APPROVED** [review]

File: openstack/objectstorage/v1/objects/requests.go (hunk 1 above)
Comment (EmilienM) [reply openstack/objectstorage/v1/objects/requests.go:577]: I think it's defined by the user in Swift config:

https://github.com/openstack/swift/blob/6d3d4197151f44bf28b51257c1a4c5d33411dcae/etc/proxy-server.conf-sample#L1029-L1034
https://github.com/openstack/swift/blob/e8cecf7fcc1630ee83b08f9a73e1e59c07f8d372/swift/common/middleware/bulk.py#L309

Comment (pierreprinetti) [reply openstack/objectstorage/v1/objects/requests.go:577]: wonderful, thank you @stephenfin @EmilienM. I have added that information in the function docs. PTAL

**Review by mandre — APPROVED** [review]

**Pull request merged**